cql3/selection/selection.cc

```diff
@@ -386,17 +388,61 @@ std::unique_ptr<result_set> result_set_builder::build() {
     return std::move(_result_set);
 }
 
-result_set_builder::restrictions_filter::restrictions_filter(::shared_ptr<restrictions::statement_restrictions> restrictions,
+namespace {
+
+using namespace expr;
+
+/// A visitor that copies an expression, but replaces token operator and unselected-column operators with the
+/// TRUE literal.
+struct drop_token_and_unselected_columns {
+    const selection& sel;
+
+    expression operator()(const conjunction& c) const {
+        auto r = c.children | std::views::transform([this] (const expression& e) { return std::visit(*this, e); });
+        return conjunction{std::vector<expression>(r.begin(), r.end())};
+    }
+
+    expression operator()(const binary_operator& binop) const {
+        return seastar::visit(binop.lhs,
+                [] (const token&) { return expression(true); },
+                [&] (const column_value& cv) {
+                    return sel.has_column(*cv.col) ? expression(binop) : expression(true);
+                },
+                [&] (auto&& default_case) { return expression(binop); });
+    }
+
+    expression operator()(bool b) const { return b; }
+};
+
+/// Compiles an expression equivalent to the given restrictions.  A filter can evaluate this expression on each
+/// row to satisfy the restrictions.
+expression compile(
+        const restrictions::statement_restrictions& restrictions,
+        const query_options& options,
+        const selection& sel) {
+    auto ex = restrictions.get_partition_key_restrictions()->expression;
+    if (restrictions.ck_restrictions_need_filtering()) {
+        ex = make_conjunction(std::move(ex), restrictions.get_clustering_columns_restrictions()->expression);
+    }
```

Comment:
Why not do it unconditionally? wouldn't rhs be true if the condition is false?

Reply:
> Why not do it unconditionally?

Because then tests fail. :)  This is equivalent to the [old code](https://sourcegraph.com/github.com/scylladb/scylla/-/blob/cql3/selection/selection.cc#L399:7&tab=references).

Reply:
But that old code is impossible to understand. I suppose I'll just have to suffer until it's gone.

Reply:
You and me both.  Sometimes I skip the rabbit hole of understanding and settle for mimicking, like here.  It's more efficient. :(